erts/preloaded/src/erlang.erl

```diff
@@ -531,15 +531,187 @@ binary_to_float(_Binary) ->
-    erlang:nif_error(undefined).
+binary_to_integer(Binary) ->
+    case erts_internal:binary_to_integer(Binary, 10) of
+        N when erlang:is_integer(N) ->
+            N;
+        [] ->
+            case big_binary_to_int(Binary, 10) of
+                N when erlang:is_integer(N) ->
+                    N;
+                Reason ->
+                    error_with_info(Reason, [Binary])
+            end;
+        badarg ->
+            badarg_with_info([Binary])
+    end.
 
 %% binary_to_integer/2
--spec binary_to_integer(Binary,Base) -> integer() when
+-spec binary_to_integer(Binary, Base) -> integer() when
       Binary :: binary(),
       Base :: 2..36.
-binary_to_integer(_Binary,_Base) ->
-    erlang:nif_error(undefined).
+binary_to_integer(Binary, Base) ->
+    case erts_internal:binary_to_integer(Binary, Base) of
+        N when erlang:is_integer(N) ->
+            N;
+        [] ->
+            case big_binary_to_int(Binary, Base) of
+                N when erlang:is_integer(N) ->
+                    N;
+                Reason ->
+                    error_with_info(Reason, [Binary,Base])
+            end;
+        badarg ->
+            badarg_with_info([Binary,Base])
+    end.
+
+big_binary_to_int(Bin, Base) ->
+    F = fun() ->
+                try big_bin_to_int(Bin, Base) of
+                    N ->
+                        erlang:exit(N)
+                catch
+                    error:Error ->
+                        erlang:exit(Error)
+                end
+        end,
+    {Pid,Ref} = erlang:spawn_monitor(F),
+    receive
+        {'DOWN',Ref,process,Pid,Result} -> Result
+    end.
+
+big_bin_to_int(Bin0, Base)
+  when erlang:is_binary(Bin0),
+       erlang:is_integer(Base), 2 =< Base, Base =< 36 ->
+    {Bin1,Sign} = get_sign(Bin0),
+    Bin = trim_zeroes(Bin1),
+    Size = erlang:byte_size(Bin),
+    if
+        Size > 4_194_304 ->
+            %% Too large even for base 2.
+            erlang:exit(system_limit);
+        Size > 1262611, Base >= 10 ->
+            erlang:exit(system_limit);
+        true ->
+            WordSize = erlang:system_info(wordsize),
+            LogRadix = digits_per_small(WordSize, Base),
+            Segments = segmentize(Bin, Size, Base, LogRadix),
+            Radix = radix(WordSize, Base),
+            Sign * combine(Segments, Radix)
+    end.
+
+segmentize(Bin, Size, Base, LogRadix) ->
+    case Size rem LogRadix of
+        0 ->
+            segmentize_1(Bin, LogRadix, Base, []);
+        NumFirst ->
+            <<First:NumFirst/binary,T/binary>> = Bin,
+            FirstInt = prim_binary_to_integer(First, Base),
+            segmentize_1(T, LogRadix, Base, [FirstInt])
+    end.
+
+segmentize_1(Bin, LogRadix, Base, Acc) ->
+    case Bin of
+        <<B:LogRadix/binary,T/binary>> ->
+            Int = prim_binary_to_integer(B, Base),
+            segmentize_1(T, LogRadix, Base, [Int|Acc]);
+        <<>> ->
+            Acc
+    end.
+
+combine(L0, Radix) ->
```

Comment:
(I see you start not from single digits but from what fits in a word. I suppose I'm asking if it's not worth doing linear algo on those "small" integers, as long as the scan isn't too long, because constants&cache.)

Reply:
The helper BIF will convert decimal numbers with up to 17 digits efficiently. We think that will take care of the majority of real-world use cases. Therefore, I have optimized the helper BIFs for correctness and simplicity.

However, after discussions with @jhogberg and @garazdawi I now think that spawning a temporary process is a bad idea and I have now eliminated that extra process. Not spawning the process will improve performance for binaries with more than 17 decimal digits.

Reply:
fwiw, i tried my proposal on some [toy code](https://gist.github.com/rgrig/91ee4ca46f85b0ae8a00050516d79396), and it's not only more complicated but also slower

> The helper BIF will convert decimal numbers with up to 17 digits efficiently. We think that will take care of the majority of real-world use cases.

I think 64b integers are a common case, especially when interacting with other languages. Probably not worth optimizing for it at this point.

> Not spawning the process will improve performance [...]

I didn't understand why there's a process spawned there, but I assumed it's my lack of Erlang knowledge. (Was it perhaps to make sure that timeouts apply to the main process even if the parsing process somwhow gets stuck?? Just a wild guess.)

Reply:
> I think 64b integers are a common case, especially when interacting with other languages. Probably not worth optimizing for it at this point.

If someone can show real-world use cases where integers whose absolute value are greater than 99999999999999999 are frequently used we could optimize this in OTP 27. 

> I didn't understand why there's a process spawned there, but I assumed it's my lack of Erlang knowledge. (Was it perhaps to make sure that timeouts apply to the main process even if the parsing process somwhow gets stuck?? Just a wild guess.)

It was an attempt at optimization (reducing garbage collection times), but after some more thinking it became clear that it in most cases it would be a [pessimization](https://en.wiktionary.org/wiki/pessimization).